server/protocols/jwt/src/main/java/org/apache/james/jwt/introspection/IntrospectionClientDefault.java

```diff
@@ -0,0 +1,138 @@
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jwt.introspection;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.apache.commons.configuration2.HierarchicalConfiguration;
+import org.apache.commons.configuration2.tree.ImmutableNode;
+import org.reactivestreams.Publisher;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Splitter;
+
+import io.netty.handler.codec.http.HttpResponseStatus;
+import reactor.core.publisher.Mono;
+import reactor.netty.ByteBufMono;
+import reactor.netty.http.client.HttpClient;
+import reactor.netty.http.client.HttpClientResponse;
+import reactor.netty.resources.ConnectionProvider;
+
+public class IntrospectionClientDefault implements IntrospectionClient {
+
+    public static class TokenIntrospectionConfiguration {
+
+        public static TokenIntrospectionConfiguration parse(HierarchicalConfiguration<ImmutableNode> configuration) throws MalformedURLException {
+            String introspectionTokenEndpoint = configuration.getString("introspectionTokenEndpoint", null);
+            Preconditions.checkNotNull(introspectionTokenEndpoint, "`introspectionTokenEndpoint` property need to be specified inside the oidc tag");
+            URL introspectionTokenEndpointURL = new URL(introspectionTokenEndpoint);
+            Map<String, Object> formAttributes = Optional.ofNullable(configuration.getString("introspectionTokenFormAttrs", null))
+                .map(TokenIntrospectionConfiguration::parseStringToMap)
+                .orElse(Map.of());
+
+            Map<String, Object> headerAttributes = Optional.ofNullable(configuration.getString("introspectionTokenHeaderAttrs", null))
+                .map(TokenIntrospectionConfiguration::parseStringToMap)
+                .orElse(Map.of());
+
+            return new TokenIntrospectionConfiguration(introspectionTokenEndpointURL, formAttributes, headerAttributes);
+        }
+
+        private static Map<String, Object> parseStringToMap(String value) {
+            return Splitter.on(",").withKeyValueSeparator("=").split(value)
+                .entrySet().stream()
+                .collect(Collectors.toMap(Map.Entry::getKey, e -> (Object) e.getValue()));
+        }
+
+        private final URL endpoint;
+        private final Map<String, Object> formAttributes;
+        private final Map<String, Object> headerAttributes;
```

Comment:
> I think form attributes should be supplied according to the payload introspected thus should not be in the configuration.

Example for keycloak. It support `token_type_hint` form. IMO, it will support similar cases

> Do we want arbitrary header configuration or are we just interested in Authorization ?

IMO, `headerAttributes` will make it more generic, if just only basic auth, we don't even need an Authorization header, we can bypass it by inline URL. (http://user:pass@domain_token_introspect)

Reply:
token_type_hint is optional no?

Reply:
> IMO, headerAttributes will make it more generic, if just only basic auth, we don't even need an Authorization header, we can bypass it by inline URL. (http://user:pass@domain_token_introspect/)

If it works it could be nice ;-)

Reply:
Yes, I tested it with the postman, and it works!
I removed `formAttributes` , `headerAttributes` to make it simpler